File: soffer/4.cpp
#include <iostream>
using namespace std;


void replaceBlank(char string[], int length) { // length 为字符数组的总长度
    if (string == NULL or length <= 0) return;
    int length_pre = 0; // 原始长度
    int length_now = 0; // 替换后长度
    int number_blank = 0; // 空格数
    int i = 0;
    while (string[i] != '\0') {
        ++length_pre;
        if (string[i] == ' ')
            ++number_blank;
        ++i;
    }
    length_now = length_pre + number_blank * 2;
    if (length_now > length) return; // 字符数组不够用
    // 从字符串'\0'开始复制
    int index_pre = length_pre;
    int index_now = length_now; // 指定两个指针
    while(index_now>index_pre) {
        if (string[index_pre] == ' ') {
            string[index_now--] = '0';
            string[index_now--] = '2';
            string[index_now--] = '%';
        } else {
            string[index_now--] = string[index_pre];
        }
        --index_pre;
    }
    cout << string << endl;
}

int main() {
    char string[100] = "I love you lee.";
    replaceBlank(string, 100);
}

File: soffer/4.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-

def replaceBlank(string):
    if string == None or len(string) < 1: 
        print('需要提供一个非空字符串')
    # 接下来用字符列表来模拟
    string = list(string)
    length_pre = len(string)
    num_blank = 0
    for char in string:
        if char == " ":
            num_blank += 1
    length_now = num_blank*2 + length_pre
    string_now = ['_' for i in range(length_now)]
    p1 = length_pre - 1
    p2 = length_now - 1
    while p1 >= 0:
        if string[p1] != " ":
            string_now[p2] = string[p1]
            p2 -= 1
        else:
            string_now[p2] = '0'
            string_now[p2-1] = '2'
            string_now[p2-2] = '%'
            p2 -= 3
        p1 -= 1
    print(''.join(string))
    print(''.join(string_now))

if __name__ == "__main__":
    s = 'I love you Lee.'
    replaceBlank(s)
